k9mail/src/main/java/com/fsck/k9/activity/ChooseSnooze.java

```diff
@@ -0,0 +1,228 @@
+import java.util.GregorianCalendar;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.TimeUnit;
+import java.text.DateFormat;
+
+
+public class ChooseSnooze extends K9ListActivity {
+    public static final String EXTRA_SNOOZE_UNTIL = "com.fsck.k9.ChooseSnooze_EXTRA_SNOOZE_UNTIL";
+
+    private static final boolean DEBUG = false;
+
+    Account mAccount;
+    MessageReference mMessageReference;
+    ArrayAdapter<SnoozeTime> mAdapter;
+
+    private DateFormat mTimeFormatter;
+    private DateFormat mTimeAndDayFormatter;
+    Resources mRes;
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setContentView(R.layout.list_content_simple);
+
+        String timeFormat;
+        if (android.text.format.DateFormat.is24HourFormat(this)) {
+            timeFormat = "EEEE HH:mm";
+        } else {
+            timeFormat = "EEEE hh:mm a";
+        }
+        mTimeAndDayFormatter = new SimpleDateFormat(timeFormat, Locale.getDefault());
+        mTimeFormatter = android.text.format.DateFormat.getTimeFormat(this);
+        mRes = getResources();
+
+        getListView().setFastScrollEnabled(true);
+        getListView().setItemsCanFocus(false);
+        getListView().setChoiceMode(ListView.CHOICE_MODE_NONE);
+        Intent intent = getIntent();
+        String accountUuid = intent.getStringExtra(ChooseFolder.EXTRA_ACCOUNT);
+        mAccount = Preferences.getPreferences(this).getAccount(accountUuid);
+        mMessageReference = intent.getParcelableExtra(ChooseFolder.EXTRA_MESSAGE);
+
+        mAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1);
+        mAdapter.addAll(getTimes());
+
+        setListAdapter(mAdapter);
+
+        getListView().setOnItemClickListener(new AdapterView.OnItemClickListener() {
+            @Override
+            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+                SnoozeTime time = mAdapter.getItem(position);
+
+                if (time.timestamp <= 0) {
+                    // choose custom time from picker
+                    showDateTimePicker(new OnDateTimePickedListener() {
+                        @Override
+                        public void onDateTimePicked(long timestamp) {
+                            finishWithResult(timestamp);
+                        }
+                    });
+                    return;
+                }
+
+                finishWithResult(time.timestamp);
+            }
+        });
+    }
+
+    private void finishWithResult(long timestamp) {
+        Intent result = new Intent();
+        result.putExtra(ChooseFolder.EXTRA_ACCOUNT, mAccount.getUuid());
+        result.putExtra(EXTRA_SNOOZE_UNTIL, timestamp);
+        result.putExtra(ChooseFolder.EXTRA_MESSAGE, mMessageReference);
+        setResult(RESULT_OK, result);
+        finish();
+    }
+
+    private List<SnoozeTime> getTimes() {
```

Comment:
Hmm, good point. You could pass the calendar in, but you'd have to reset the calendar to now every time or the behaviour would be undefined. Let's just leave it as is and moan at Java's DateTime API for being lame.